File: LLD/ObserverDesignPattern/Java/ObserverDesignPattern.java

import java.util.ArrayList;
import java.util.List;

interface  ISubscriber {
    void update();
}

interface IChannel {
    void subscribe(ISubscriber sub);
    void unSubscribe(ISubscriber sub);
    void notifySubscribers();
}

class Channel implements IChannel {
    private  String channelName;
    private String lastVideo;
    private List<ISubscriber> subscribers;

     Channel(String channelName) {
        this.channelName = channelName;
        this.subscribers = new ArrayList<>();
    }
    @Override
    public void subscribe(ISubscriber sub){
        if(sub != null && !subscribers.contains(sub)) {
            subscribers.add(sub);
            System.out.println(sub + " subscribed to " + channelName);
        }
    }

    @Override
    public void unSubscribe(ISubscriber sub){
        if(sub != null && subscribers.contains(sub)) {
            subscribers.remove(sub);
            System.out.println(sub + " unsubscribed from " + channelName);
        }
    }

    @Override
    public void notifySubscribers(){
        for(ISubscriber sub : subscribers) {
            sub.update();
        }
    }

    public void uploadVideo(String videoTitle) {
        this.lastVideo = videoTitle;
        System.out.println("New video uploaded: " + videoTitle + " on " + channelName);
        notifySubscribers();
    }

    public String getVideo(){
        return lastVideo;
    }

}

class Subscriber implements ISubscriber {
    private String  name;
    private Channel channel;

    public Subscriber(String name, Channel channel) {
        this.name = name;
        this.channel = channel;
    }

    @Override
    public String toString() {
        return name;
    }

    
    @Override
    public void update() {
        System.out.println(name + " has been notified of a new video: " + channel.getVideo());
    }
}

public class ObserverDesignPattern {
    public static void main(String[] args) {
        Channel techChannel = new Channel("Tech Insights");
        Subscriber alice = new Subscriber("Alice", techChannel);
        Subscriber bob = new Subscriber("Bob", techChannel);

        techChannel.subscribe(alice);
        techChannel.subscribe(bob);

        techChannel.uploadVideo("Understanding Design Patterns in Java");

        techChannel.unSubscribe(bob);
        techChannel.uploadVideo("Advanced Java Programming Techniques");
    }
}

File: LLD/ObserverDesignPattern/C++/ObserverDesignPattern.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class ISubscriber {
    public:
        virtual void update() = 0;
        virtual ~ISubscriber(){};
};
class IChannel {
    public:
        virtual void subscribe(ISubscriber* sub) = 0;
        virtual void unSubscribe(ISubscriber* sub) = 0;
        virtual void notify() = 0;

        virtual ~IChannel(){};
};

class Channel : public IChannel {
    private:
        string name;
        string lastVideo;
        vector<ISubscriber*>subscribers;

    public:
        Channel(const string& name) : name(name){};

        void subscribe(ISubscriber* sub) override {
            if(find(subscribers.begin(), subscribers.end(), sub) == subscribers.end()) {
                subscribers.push_back(sub);
            }
        }

        void unSubscribe(ISubscriber* sub) override{
            auto it = find(subscribers.begin(),subscribers.end(),sub);
            if(it!= subscribers.end()){
                subscribers.erase(it);
            }
        }
        void notify() override {
            for(auto& sub : subscribers) {
                sub->update();
            }
        }
        void uploadVideo(const string& video) {
            lastVideo = video;
            cout << "Channel: " << name << " uploaded a new video: " << video << endl;
            notify();
        }

        string getVideo() {
            return lastVideo;
        }

        string getName() {
            return name;
        }

};

class Subscriber : public ISubscriber {
    private:
        string name;
        Channel* channel;
    public:

    Subscriber( const string &name,Channel *channel) : name(name),channel(channel){};
    void update() override{
        cout  << "👋 Hey " << name << " New video uploaded on channel: " << channel->getName() <<" , "<<channel->getVideo() <<endl;
    }
};

int main() {
    Channel* channel = new Channel("CoderWorld");

    Subscriber* subs1 = new Subscriber("Ram", channel);
    Subscriber* subs2 = new Subscriber("Alice", channel);

    channel->subscribe(subs1);
    channel->subscribe(subs2);

    channel->uploadVideo("Observer Pattern Tutorial");

    channel->unSubscribe(subs1);

    channel->uploadVideo("Observer Patter Notes");

    // Clean up
    delete subs1;
    delete subs2;
    delete channel;
    return 0;
}


/* OUTPUT:
 Channel: CoderWorld uploaded a new video: Observer Pattern Tutorial
👋 Hey Ram New video uploaded on channel: CoderWorld , Observer Pattern Tutorial
👋 Hey Syam New video uploaded on channel: CoderWorld , Observer Pattern Tutorial
Channel: CoderWorld uploaded a new video: Observer Patter Notes
👋 Hey Syam New video uploaded on channel: CoderWorld , Observer Patter Notes
 */